Bug 1112282: fix javascript errors in 2011 annual report

raebex opened the pull request.

craigcook commented on bedrock/foundation/templates/foundation/annualreport/2011.html:
```diff
@@ -696,164 +696,167 @@ <h2 class="section-title">{{ _('Community') }}</h2>
-
+        <div class="jcarousel-container">
+          <button type="button" class="btn-prev">Previous</button>
+          <button type="button" class="btn-next">Next</button>
```
These strings should be wrapped for l10n, <code>{{ _('Previous') }}</code> and <code>{{ _('Next') }}</code> (I think we may have these already translated elsewhere).

raebex commented:
Totally slipped my mind...thanks for the catch! I made another commit with the changes.

craigcook commented:
This isn't necessarily a blocker, but when I change the viewport from large to small or small to large, the slideshow layout breaks until I reload the page. Granted, this isn't a common scenario for normal people but one realistic use case is a small tablet going from portrait to landscape and vice versa. That was the reason for the <code>$window.resize</code> bit so maybe it's worth bringing that back. Perhaps @jpetto or @alexgibson have additional thoughts? Is this worth worrying about?

craigcook commented:
The thumbnail list still has its default bullets:
![thumb-dots](https://cloud.githubusercontent.com/assets/205591/5866092/12fdf114-a243-11e4-8dba-c0c1475188ba.png)
Just needs a <code>list-style: none;</code>

alexgibson commented:
@craigcook - looks like there is a matchMedia event listener here to handle the resize changes? 

I haven't tested, but maybe the layout needs s little adjusting on resize? Also good to make sure to include the polyfill `'js/libs/matchMedia.addListener.js'` for IE9 using a conditional comment.

craigcook commented:
Yeah, it's using a matchMedia listener but it doesn't seem to be firing in my release Firefox. But now I've tried a few other browsers and it works as it should in Chrome, Safari, and even DevEdition so maybe it's just me?

And good catch on adding the IE9 polyfill, I haven't tested IE yet.

alexgibson commented:
@craigcook - just tested 35.0 and `matchMedia` seems to be firing OK for me?

craigcook commented:
Really weird... I did a minimal test in a jsfiddle and matchmedia fires as expected, it's just not working for me on this page in Fx35. Even tried with a clean profile. empty cache, and no addons and still doesn't work. I'm baffled. But if we can't duplicate it elsewhere then I'm happy to call it a fluke.

@raebex for the IE9 polyfill, you can refer to https://github.com/mozilla/bedrock/blob/master/bedrock/mozorg/templates/mozorg/home/home-new.html#L34 and https://github.com/mozilla/bedrock/blob/master/bedrock/settings/base.py#L908 for an example where we've done this elsewhere. Let me know if you hit any snags.

alexgibson commented:
Strange, maybe worth getting either @jpetto or @sgarrity to test?

raebex commented:
I'll revisit the window resizing javascript, or it might even be something with the CSS positioning. I've had problems like this with absolute positioning in the past. Thanks @alexgibson and @craigcook for the IE9 polyfill reference. I should have an update by tonight!

jpetto commented:
Just gave it a test and found a layout bug going from desktop to mobile. Starting on desktop, click the next button/arrow on the slider a couple times, then shrink down to mobile. You'll see the list of thumbnails has a negative top offset:

![screen shot 2015-01-23 at 2 57 43 pm](https://cloud.githubusercontent.com/assets/317532/5882586/7a1ccece-a310-11e4-94eb-67438cd49763.png)

A quick inspect element shows a `top: -117px;` inline style on the `ul#story-slider`. May be as easy as removing inline styles on that element when `matchMedia` detects mobile viewports.

This was the only bug I noticed, and was consistent in both Fx & Chrome.

raebex commented:
@jpetto I just added ```top: 0; ``` to ```#story-slider``` and it seems to get rid of the layout problem, at least for me. There weren't any inline styles on ```#story-slider``` so I don't know why the browser was adding those.

@alexgibson @craigcook is this the right way to reference the matchMedia polyfill?
```
{% block js %}
  {{ js('annual_2011') }}
  <!--[if IE 9]>
    <script type="text/javascript" src="{{ MEDIA_URL }}js/libs/matchMedia.addListener.js"></script>
  <![endif]-->
{% endblock %}
```

Or should I make 'annual_2011_ie9' in settings/base.py and include the file there, then reference it like  ```{{ js('annual_2011_ie9') }}```

craigcook commented:
Let's go with  <code>{{ js('annual_2011_ie9') }}</code>.

Also make sure the IE9 script comes *before* the general annual_2011 script so IE9 loads the polyfill first.

jpetto commented:
@raebex - It looks like the carousel script is adding inline styles when clicking the next/prev arrows.

https://cloudup.com/cki0X26ixii

Are you seeing the same behavior?

You'll probably need to do something like `$('#story-slider').attr('style', '');` when squishing down to mobile size.

jpetto commented:
For a little more detail, it looks like the carousel script (or something else?) is altering inline styles when going from mobile to desktop and back. Weird!

https://cloudup.com/chle9BZa7oD

raebex commented:
I made an update! I tested in IE 9 and have no script errors. For some reason no matter how many times I put ```$('#story-slider').attr('style', '');``` or ```$('#story-slider').removeAttr('style');``` in the javascript, inline styles still showed up. Though having ```top: 0;``` in the CSS at least makes it look correct.

jpetto commented on media/js/foundation/annual2011.js:
```diff
@@ -14,39 +14,34 @@
         $("#video-stage").show();
-    } else {
+    }
+    else {
```
Nit pick - put on same line as `}` above.

jpetto commented on media/js/foundation/annual2011.js:
```diff
@@ -284,7 +285,7 @@
     // Remove the carousel when we don't need it, redo the thumbnails and show the original
     // This is for smaller viewports that don't get the slider
     function removeCarousel() {
```
Doesn't look like this function is being called any more. Do we still need it?

jpetto commented on media/js/foundation/annual2011.js:
```diff
@@ -215,29 +210,21 @@
-            itemLastInCallback: { onAfterAnimation: disableButtons },
-            initCallback: controlButtons
-        });
+        $sliderPrime.removeAttr("style");
```
Hm, I don't think you want to make this call here - it seems it should be called when moving into a mobile sized viewport. Probably in the `matchMedia` listener above?

jpetto commented:
I'm still seeing the layout bug when going from desktop to mobile. :frowning:

@raebex - Can you replicate? Starting at desktop view, click the right arrow in the carousel a few times, then resize down to mobile. Do you see the thumbnails with a negative top value?

I peeked at the lib we're using and saw that there's a [`destroy` method available](http://sorgalla.com/jcarousel/docs/reference/api.html#destroy).

I did a little tinkering locally, and found that using the `destroy` method along with a listener removes the inline styles.

Inside `setupCarousel`, I changed the initializer to take custom action when the carousel is destroyed:

```
$('.jcarousel-clip').jcarousel().on('jcarousel:destroyend', function() {
    $sliderPrime.attr('style', '');
    console.log('offsets should be removed');
    console.log($sliderPrime.attr('style'));
});
```

Then, in the `matchMedia` listener for mobile screens, I destroy the carousel (which fires the above listener): `$('.jcarousel-clip').jcarousel('destroy');`

I pushed this code to [a branch on my fork](https://github.com/jpetto/bedrock/blob/bug-1112282-2011-annual-report-js-errors/media/js/foundation/annual2011.js) if you want to check it out in context.

This seems to solve the problem for me, but the code could probably be cleaned up a little.

That might be a lot of information to take in. Let me know (in #www on IRC?) if you have questions. :smile:

raebex commented:
@jpetto I implemented your code but inline styles were still being added for me. I kept the code in, but I also moved ```#story-slider { position: relative; }``` to the tablet/desktop media query and it seems to look right now.

jpetto commented on media/js/foundation/annual2011.js:
```diff
@@ -215,12 +209,10 @@
-        $sliderPrime.hide();
-        $("#story-slider-clone").jcarousel({
+        $('.jcarousel-clip').jcarousel().on('jcarousel:destroyend', function() {
+            $sliderPrime.attr('style', '');
```
Ah, I think this isn't removing the styles because the `.jcarousel()` call immediately following overwrites these settings. The way to make it work would be adding the `.on('jcoursel:destroyend', ...` bit on to the initialization call below. But let's not do that!

I'd say you can wipe out this bit and we'll just use your CSS solution.

jpetto commented:
@raebex - I think your CSS solution is better than doing heavy lifting in JS. All seems well in IE 9+, which I'd say is good enough for this page. It's at least readable in IE 8. Will confirm with the team on Monday and hopefully push then.

In the mean time, can you squash the commits?

Thanks! :smile:

raebex commented:
@jpetto sounds good! I squashed and things are ready to go. Thanks so much for your help! :D

jpetto commented:
Looking good to me. r+

Thanks for all the work @raebex! :hatched_chick:
